content/blog/2024-09-10-memory-access-amortization.md

```diff
@@ -0,0 +1,158 @@
+# Aim for 2 short sentences.
+description= "Improve performance by enabling the processor's speculative execution capability"
+# 'authors' are the folks who wrote or contributed to the post.
+# Each author corresponds to a biography file (more info later in this document)
+authors= [ "dantouitou", "uriyagelnik"]
++++
+<style>
+  .inverted-code {
+    font-family: "Courier New", Courier, monospace;
+    background-color: #333; /* Dark background */
+    color: #f0f0f0;         /* Light text */
+    padding: 10px;
+    border-radius: 5px;
+    display: block;
+    white-space: pre-wrap;
+  }
+</style>
+
+ In the [first part](https://valkey.io/blog/unlock-one-million-rps/) of this blog, we described how we offloaded almost all I/O operations to I/O threads, thereby freeing more CPU cycles in the main thread to execute commands. When we profiled the execution of the main thread, we found that a considerable amount of time was spent waiting for external memory. This was not entirely surprising, as when accessing random keys, the probability of finding the key in one of the processor caches is relatively low.  Considering that external memory access latency is approximately 50 times higher than L1 cache, it became clear that despite showing 100% CPU utilization, the main process was mostly “waiting”. In this blog, we describe the technique we have been using to increase the number of parallel memory accesses, thereby reducing the impact that external memory latency has on performance.
+
+### Speculative execution and linked lists 
+Speculative execution is a performance optimization technique used by modern processors, where the processor guesses the outcome of conditional operations and executes in parallel subsequent instructions ahead of time. Dynamic data structures, such as linked lists and search trees, have many advantages over static data structures: they are economical in memory consumption, provide fast insertion and deletion mechanisms, and can be resized efficiently. However, some dynamic data structures have a major drawback: they hinder the processor's ability to speculate on future memory load instructions that could be executed in parallel. This lack of concurrency is especially problematic in very large dynamic data structures, where most pointer accesses result in high-latency external memory access.
+
+In this blog, Memory Access Amortization, a method that facilitates speculative execution to improve performance, is introduced along with how it is applied in Valkey. The basic idea behind the method is that by interleaving the execution of operations that access random memory locations, one can achieve significantly better performance than by executing them serially.
+
+To depict the problem we are trying to solve consider the following [function](/assets/C/list_array.c) which gets an array of linked list and returns sum of all values in the lists:
+<pre class="inverted-code">
```

Comment:
Use triple backticks tagged with "```c" instead of `pre`

<img width="537" alt="Screenshot 2024-09-11 at 1 41 54 PM" src="https://github.com/user-attachments/assets/9ff19abf-3dc0-4726-b5b5-20aa092cc4a0">


You get a nicer syntax highlight

![Screenshot 2024-09-11 at 1 41 58 PM](https://github.com/user-attachments/assets/e5e448b0-8465-4312-aa32-259e9c3e4bb0)